### PR TITLE
Add squash merges option for editorial workflow

### LINKED ISSUE #1328: Use 'squash and merge' if possible when publishing posts
**I mentioned this in the Gitter chat and was told to write up an issue on this.**

### Current behavior:

When the CMS publishes a post, it merges the post behind the scenes. However the resulting commit history shows every single minor update done to the post, it keeps the entire commit history.

### Suggested change in behavior:

The CMS should use the 'Squash and Merge' functionality instead when available.

### Reasoning:

This would make sense for the CMS because when you publish a post, most people (at least I hope so) don't want to see the commit history of the post, it just creates a bunch of visual noise in the git log/commit history.

### PR DESCRIPTION
**Pull request by delucis**
### Summary
Closes #1328

Exposes a `backend.squash_merges` option, which allows users to configure the merge method used when merging Github pull requests. The standard behaviour remains unchanged, but setting `squash_merges: true` will use the `squash` merge method instead (see [Github API docs](https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button)).

#### Why?
When using the Editorial Workflow, every small change generates at least two commits in the repository history with standard merges, more if content was edited several times before publishing. For content-heavy sites, this can mean the commit history is overloaded with (pretty uninformative) merge commits. Squash merges allow users to minimise clutter in the commit history if they want to.

### Considerations
I’m not sure how this fits together with planning for other backends.

### Test plan
- `npm run test` passed
- `npm run lint` threw `281935 problems (245884 errors, 36051 warnings)`, which I’m hoping my changes didn’t cause 😂 
- Tested in a branch of a Nuxt + Netlify CMS site deployed to Netlify (using `git-gateway`) and seems to be working, would be happy to follow up suggestions for further testing

### Description for the changelog

Add squash merges option for editorial workflow

### A picture of a cute animal (not mandatory but encouraged)

An okapi 😀 

An Okapi standing in long grass

### CODE REVIEW

**Comment by verythorough** [comment]
Deploy preview for *netlify-cms-www* ready!

Built with commit c6ab1768153fe53805b2afe32243538e35472e7a

https://deploy-preview-1330--netlify-cms-www.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit c6ab1768153fe53805b2afe32243538e35472e7a

https://deploy-preview-1330--cms-demo.netlify.com

**Comment by verythorough** [comment]
Deploy preview for *cms-demo* ready!

Built with commit 94964b8ea1c9e7abcd763afc022f931c524b04aa

https://deploy-preview-1330--cms-demo.netlify.com

**Comment by Benaiah** [comment]
@delucis re: fitting in with other backends: part of the backend API discussion has been around separating options that are specific to a given backend from options that would apply to any backend used by the CMS<sup>1</sup>. This is currently unimplemented, so I think the current location is fine, but I think we may want to document this as a preview feature so that we're not committed to maintaining the current key location when it may change soon.

So basically, there are some considerations around other backends, but I don't think there's any related changes needed in this PR before it's merged.

<sup>1</sup> e.g., something like the following:
```yaml
backend:
  name: github
  settings:
    squash_merges: true # this setting is github-specific
```
It probably won't look exactly like this, but that's the general idea.

**Comment by delucis** [comment]
@Benaiah Thanks! If this should be documented as a “preview feature”, should I document it on the [“Beta Features!”](https://www.netlifycms.org/docs/beta-features/) page? Or just add a note [where it is at the moment](https://deploy-preview-1330--netlify-cms-www.netlify.com/docs/authentication-backends/#options)?

<hr/>

Out of curiosity, just looked into how squash merging would work with the BitBucket and GitLab APIs, so posting that here for future reference:

- BitBucket pull requests can be squash merged by passing `merge_strategy: squash` when merging [[Docs →](https://developer.atlassian.com/bitbucket/api/2/reference/resource/repositories/%7Busername%7D/%7Brepo_slug%7D/pullrequests/%7Bpull_request_id%7D/merge)]
- GitLab merge requests have a boolean `squash` property that is set directly on the request (at create/update) rather than when the request is merged [[Docs →](https://docs.gitlab.com/ee/api/merge_requests.html#create-mr)]

**Comment by Benaiah** [comment]
@delucis I think the Beta Features page would be the best spot (@erquhart @verythorough what do you think?). Thanks for looking into GitLab and BitBucket! It looks like it should be straightforward to support this feature on those backends as well in the future.

**Comment by delucis** [comment]
OK! Moved the documentation to the Beat Features page: https://deploy-preview-1330--netlify-cms-www.netlify.com/docs/beta-features/#squash-merge-github-pull-requests

File: website/site/content/docs/beta-features.md [hunk 1]
```diff
@@ -77,3 +77,10 @@ import styles from '!css-loader!sass-loader!../main.scss'
 CMS.registerPreviewStyle(styles.toString(), { raw: true })
 ```
 
+## Squash merge GitHub pull requests
+When using the [Editorial Workflow publish mode](/docs/configuration-options/#publish-mode) with the `github` or `git-gateway` backends, Netlify CMS creates pull requests to hold drafts while they are being edited and then merges them into your main branch when you press publish. By default, these are merged preserving the history of all the individual commits (i.e. changes where you pressed save). If instead you would like [“squash”](https://help.github.com/articles/about-pull-request-merges/#squash-and-merge-your-pull-request-commits) these commits, reducing clutter in your repository’s history, you can set the following option in your `config.yml`:
```
Comment (verythorough) [inline website/site/content/docs/beta-features.md]: Great description! The one suggestion I would make is to simplify the second-to-last sentence. Parentheticals and Latin abbreviations like _i.e._ can be difficult for translators, etc. I suggest ending the sentence after "individual commits" and changing the parenthetical to a sentence, like:
> In other words, you'll have a commit in your history for every time you pressed the save button.

I would also suggest adding a paragraph break before the last sentence ("If instead...").

**Comment by delucis** [comment]
@verythorough Thanks for the review! Definitely an improvement: https://deploy-preview-1330--netlify-cms-www.netlify.com/docs/beta-features/#squash-merge-github-pull-requests

**Comment by delucis** [comment]
@Benaiah @verythorough Let me know if there’s anything I can do to improve this before it’s merged!

File: src/actions/config.js [hunk 2]
```diff
@@ -50,6 +50,9 @@ export function validateConfig(config) {
   if (typeof config.getIn(['backend', 'name']) !== 'string') {
     throw new Error("Error in configuration file: Your `backend.name` must be a string. Check your config.yml file.");
   }
+  if (!isBoolean(config.getIn(['backend', 'squash_merges'], false))) {
```
Comment (erquhart) [inline src/actions/config.js]: Leave this out, trying to avoid validation of beta features, plus this is specific to one backend. After 2.0 I expect we'll provide a way for specific backends to validate their configuration, as a similar effort is underway for widgets.

**Comment by delucis** [comment]
@erquhart Thanks! Removed that block.

**Comment by erquhart** [comment]
Thanks again @delucis!
